src/data/nat/basic.lean

```diff
@@ -1062,8 +1069,34 @@ lemma with_bot.add_eq_one_iff : ∀ {n m : with_bot ℕ}, n + m = 1 ↔ (n = 0 
 by apply nat.less_than_or_equal.rec h0; exact h1
 
 @[elab_as_eliminator]
-lemma decreasing_induction {P : ℕ → Prop} (h : ∀n, P (n+1) → P n) {m n : ℕ} (nm : m ≤ n) (hP : P n) :
-  P m :=
-by { induction nm with n nm ih, exact hP, exact ih (h _ hP) }
+def decreasing_induction {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {m n : ℕ} (mn : m ≤ n)
+  (hP : P n) : P m :=
+le_rec_on mn (λ k ih hsk, ih $ h k hsk) (λ h, h) hP
+
+lemma decreasing_induction_self {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {n : ℕ} (nn : n ≤ n)
+  (hP : P n) : (decreasing_induction h nn hP : P n) = hP :=
+by { dunfold decreasing_induction, rw [le_rec_on_self] }
+
+lemma decreasing_induction_succ {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {m n : ℕ} (mn : m ≤ n)
+  (msn : m ≤ n + 1) (hP : P (n+1)) :
+  (decreasing_induction h msn hP : P m) = decreasing_induction h mn (h n hP) :=
+by { dunfold decreasing_induction, rw [le_rec_on_succ] }
+
+lemma decreasing_induction_succ' {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {m : ℕ} (msm : m ≤ m + 1)
+  (hP : P (m+1)) : (decreasing_induction h msm hP : P m) = h m hP :=
+by { dunfold decreasing_induction, rw [le_rec_on_succ'] }
+
+lemma decreasing_induction_trans {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {m n k : ℕ}
+  (mn : m ≤ n) (nk : n ≤ k) (hP : P k) :
+  (decreasing_induction h (le_trans mn nk) hP : P m) =
+  decreasing_induction h mn (decreasing_induction h nk hP) :=
+by { induction nk with k nk ih, rw [decreasing_induction_self],
+     rw [decreasing_induction_succ h (le_trans mn nk), ih, decreasing_induction_succ] }
+
+lemma decreasing_induction_succ_left {P : ℕ → Sort*} (h : ∀n, P (n+1) → P n) {m n : ℕ}
+  (smn : m + 1 ≤ n) (mn : m ≤ n) (hP : P n) :
```

Comment:
`simp`

Reply:
RHS contains a proof `smn` that cannot be inferred from LHS.